Lib/test/test_capi/test_file.py

```diff
@@ -1,22 +1,67 @@
 import os
 import unittest
-from test.support import import_helper
+from test import support
+from test.support import import_helper, os_helper
 
 _testcapi = import_helper.import_module('_testcapi')
 
 
 class CAPIFileTest(unittest.TestCase):
     def test_py_fopen(self):
+        # Test Py_fopen() and Py_fclose()
+        class FSPath:
+            def __init__(self, path):
+                self.path = path
+            def __fspath__(self):
+                return self.path
+
+        with open(__file__, "rb") as fp:
+            source = fp.read()
+
         for filename in (__file__, os.fsencode(__file__)):
             with self.subTest(filename=filename):
-                content = _testcapi.py_fopen(filename, "rb")
-                with open(filename, "rb") as fp:
-                    self.assertEqual(fp.read(256), content)
+                data = _testcapi.py_fopen(filename, "rb")
+                self.assertEqual(data, source[:256])
+
+                data = _testcapi.py_fopen(FSPath(filename), "rb")
+                self.assertEqual(data, source[:256])
+
+        for filename in (
+            os_helper.TESTFN,
+            os.fsencode(os_helper.TESTFN),
+            os_helper.TESTFN_UNDECODABLE,
+            os_helper.TESTFN_UNENCODABLE,
```

Comment:
TESTFN_UNDECODABLE and TESTFN_UNENCODABLE can be None. You should not test in that case. If TESTFN_UNDECODABLE is not None, test also os.fsdecode(TESTFN_UNDECODABLE).

Reply:
It's not possible to create a file named TESTFN_UNDECODABLE on macOS/WASI, so I removed TESTFN_UNDECODABLE test.

I added a test to check if TESTFN_UNENCODABLE is None.